src/data_source/storage/sql.rs

```diff
@@ -1170,6 +1192,401 @@ where
+        // Hashmap will be used to get the hash value of the nodes children and the node itself.
+        let hashes_query = self
+            .query(
+                "SELECT * FROM hash WHERE id = ANY( $1)",
+                [sql_param(&hash_ids.into_iter().collect::<Vec<i32>>())],
+            )
+            .await?;
+
+        let hashes: HashMap<_, _> = hashes_query
+            .map(|row| HashTableRow::try_from(row?).map(|h| (h.id, h.value)))
+            .try_collect()
+            .await?;
+
+        let mut proof_path = Vec::new();
+        for Node {
+            hash_id,
+            children,
+            children_bitvec,
+            index,
+            entry,
+            ..
+        } in nodes.iter()
+        {
+            {
+                let value = hashes.get(hash_id).ok_or(QueryError::NotFound)?;
+                if let (Some(children), Some(children_bitvec)) = (children, children_bitvec) {
+                    let mut children = children.iter();
+
+                    // Reconstruct the Children MerkleNodes from storage.
+                    // Children bit_vec is used to create forgotten  or empty node
+                    let child_nodes = children_bitvec
+                        .iter()
+                        .map(|bit| {
+                            if bit {
+                                let value = hashes
+                                    .get(children.next().ok_or(QueryError::NotFound)?)
+                                    .ok_or(QueryError::NotFound)?;
+                                Ok(Arc::new(MerkleNode::ForgettenSubtree {
+                                    value: T::deserialize_compressed(value.as_slice())
+                                        .map_err(ParseError::Deserialize)?,
+                                }))
+                            } else {
+                                Ok(Arc::new(MerkleNode::Empty))
+                            }
+                        })
+                        .collect::<QueryResult<Vec<_>>>()?;
+                    // Use the Children merkle nodes to reconstruct the branch node
+                    proof_path.push(MerkleNode::Branch {
+                        value: T::deserialize_compressed(value.as_slice())
+                            .map_err(ParseError::Deserialize)?,
+                        children: child_nodes,
+                    });
+                } else {
+                    // If the children for a node does not exist then the node is actually a leaf node
+                    proof_path.push(MerkleNode::Leaf {
+                        value: T::deserialize_compressed(value.as_slice())
+                            .map_err(ParseError::Deserialize)?,
+                        pos: serde_json::from_value(index.clone()).map_err(ParseError::Serde)?,
+                        elem: serde_json::from_value(entry.clone()).map_err(ParseError::Serde)?,
+                    });
+                }
+            }
+        }
+
+        Ok(proof_path)
+    }
```

Comment:
We're missing the check that data isn't missing: https://www.notion.so/espressosys/Efficient-storage-and-queries-for-Merklized-state-8a8ccd79b73d4a7abe9a8df4208983dd?pvs=4#6a1b66321c47486bb318bd14d78afc29